ntp-os-clock/src/unix.rs

```diff
@@ -0,0 +1,402 @@
+    jitter: 0,
+    shift: 0,
+    stabil: 0,
+    jitcnt: 0,
+    calcnt: 0,
+    errcnt: 0,
+    stbcnt: 0,
+    tai: 0,
+    __unused1: 0,
+    __unused2: 0,
+    __unused3: 0,
+    __unused4: 0,
+    __unused5: 0,
+    __unused6: 0,
+    __unused7: 0,
+    __unused8: 0,
+    __unused9: 0,
+    __unused10: 0,
+    __unused11: 0,
+};
+
+#[cfg(any(target_os = "freebsd", target_os = "macos"))]
+pub(crate) const EMPTY_TIMEX: libc::timex = libc::timex {
+    modes: 0,
+    offset: 0,
+    freq: 0,
+    maxerror: 0,
+    esterror: 0,
+    status: 0,
+    constant: 0,
+    precision: 0,
+    tolerance: 0,
+    ppsfreq: 0,
+    jitter: 0,
+    shift: 0,
+    stabil: 0,
+    jitcnt: 0,
+    calcnt: 0,
+    errcnt: 0,
+    stbcnt: 0,
+};
+
+/// NTP Clock that uses the unix NTP KAPI clock functions to get/modify the
+/// current time.
+// Implementation note: this is intentionally a bare struct, the NTP Clock defined
+// in the NTP KAPI is unique and no state is needed to interact with it.
+#[derive(Debug, Default, Clone)]
+pub struct UnixNtpClock {
+    clock: libc::clockid_t,
+}
+
+impl UnixNtpClock {
+    pub fn realtime() -> Self {
+        Self::custom(libc::CLOCK_REALTIME)
+    }
+
+    pub fn custom(id: libc::clockid_t) -> Self {
+        Self { clock: id }
+    }
+
+    #[cfg_attr(target_os = "linux", allow(unused))]
+    pub fn ptp0() -> Result<Self, Error> {
+        let path = CStr::from_bytes_with_nul(b"/dev/ptp0\0").unwrap();
+        // SAFETY
+        //
+        // we know this is a file descriptor of a clock
+        unsafe { Self::from_path(path) }
+    }
+
+    unsafe fn from_path(path: &CStr) -> Result<Self, Error> {
+        let fd = match unsafe { libc::open(path.as_ptr(), libc::O_RDWR) } {
+            -1 => return Err(convert_errno()),
+            valid => valid,
+        };
+
+        Ok(unsafe { Self::from_file_descriptor(fd as RawFd) })
+    }
+
+    unsafe fn from_file_descriptor(fd: RawFd) -> Self {
+        let id = ((!(fd as libc::clockid_t)) << 3) | 0b11;
+
+        Self::custom(id)
+    }
+
+    fn clock_gettime(&self) -> Result<libc::timespec, Error> {
+        let mut timespec = libc::timespec {
+            tv_sec: 0,
+            tv_nsec: 0,
+        };
+
+        cerr(unsafe { libc::clock_gettime(self.clock, &mut timespec) })?;
+
+        Ok(timespec)
+    }
+
+    fn clock_settime(&self, mut timespec: libc::timespec) -> Result<(), Error> {
+        while timespec.tv_nsec > 1_000_000_000 {
+            timespec.tv_sec += 1;
+            timespec.tv_nsec -= 1_000_000_000;
+        }
+
+        unsafe { cerr(libc::clock_settime(self.clock, &timespec))? };
+
+        Ok(())
+    }
+
+    fn clock_adjtime(&self, timex: &mut libc::timex) -> Result<(), Error> {
+        // We don't care about the time status, so the non-error
+        // information in the return value of ntp_adjtime can be ignored.
+        // The ntp_adjtime call is safe because the reference always
+        // points to a valid libc::timex.
```

Comment:
This comment is out of date, plus are there requirements here on self.clock?

Reply:
updated. There are no requirements on `self.clock`; an invalid clock is handled gracefully with an EINVAL IO error